src/Text/Pandoc/Readers/Markdown.hs

```diff
@@ -1819,30 +1819,48 @@ bracketedSpan = do
   try $ do
     (lab,_) <- reference
     attr <- attributes
-    return $ if isSmallCaps attr
-                then B.smallcaps <$> lab
-                else if isUnderline attr
-                        then B.underline <$> lab
-                        else B.spanWith attr <$> lab
-
--- | We treat a span as SmallCaps if class is "smallcaps" (and
--- no other attributes are set or if style is "font-variant:small-caps"
--- (and no other attributes are set).
+    return $ case smallCapsAttr attr of
+      Just ("", [], []) -> B.smallcaps <$> lab
+      Just scAttr       -> B.spanWith scAttr . B.smallcaps <$> lab
+      Nothing           -> case underlineAttr attr of
+        Just ("", [], []) -> B.underline <$> lab
+        Just ulAttr       -> B.spanWith ulAttr . B.underline <$> lab
+        Nothing           -> B.spanWith attr <$> lab
+
```

Comment:
Maybe, to be consistent with the recent change to the HTML writer, we should allow `[test]{.underline .smallcaps}` to turn into `Underline [SmallCaps [Str "test"]]`?  Not sure, what do you think?

Reply:
This could be done fairly elegantly with a fold over the classes, similarly to what I did in the HTML writer.

Reply:
Nice! Done.